File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
public:
    int findmaxi(vector<int>&piles){
        int maxi=INT_MIN;
        for(int i=0;i<piles.size();i++){
            maxi=max(maxi,piles[i]);
        }
        return maxi;
    }
    long long totalhours(vector<int>& piles,int mid){
        long long th=0;
        int n=piles.size();
        for(int i=0;i<n;i++){
            th+=ceil((double)piles[i] /(double)mid);
        }
        return th;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int low=1;
        int high=findmaxi(piles);
        while(low<=high){
            int mid= (high + low) / 2;
            long long t_h=totalhours(piles,mid);
            if(t_h<=h){
                high=mid-1;
            }
            else{
                low=mid+1;
            } 
        }
        return low;

    }

};

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        Arrays.sort(piles);
        int low=1;
        int high=piles[piles.length-1];
        int ans=0;
        long k=0;
        while(low<=high){
            int mid=(low+high)/2;
            k=hours(piles,mid);
            System.out.println(mid+" "+k);
            if(k>h){
                low=mid+1;
            }
            else{
                ans=mid;
                high=mid-1;
            }
        }
        return ans;
    }

    public long hours(int[] piles,int k){
        long h=0;
        for(int pile:piles){
            h += (pile + k - 1) / k;
        }
        return h;
    }
}